### DISCUSSION TITLE
Hex editor preview panel scrolling

### DISCUSSION THREAD

**Discussion by SlothLady** [WerWolv/ImHex#1163] [Feedback]
Have only just started using the program and it's been very helpful, just noticed though that when scrolling in the hex editor view, the lines slowly shift up and down depending on which direction you scroll, I'm not sure if that's intended but I just thought I'd point it out, Thank you :)

**Comment by paxcut** [comment]
Can you please elaborate a bit into describing the unwanted or unexpected behavior? You describe lines moving when you scroll but that is the way scrolling affects the display of the file. Do the lines continue to move when you stop scrolling? Is the issue that they move slowly? Are they moving in the opposite direction of what they should? sorry if i am missing something obvious.

**Comment by SlothLady** [comment]
Sorry I might have phrased this really poorly. When scrolling, I would expect the lines to stay in the same place if that makes sense? Like the data should move but the top and bottom of each line wouldn't move on the page however when scrolling the top and bottom of each line moves by a few pixels as you scroll up and down. I hope that explains it better, sorry I can't really think of any other way to describe it haha.

Thank you

**Comment by paxcut** [comment]
I think this is what you mean but feel free to tell me otherwise. 

1) Start with a file that doesnt fit entirely in the vertical space the editor has. This means it must have a vertical scrollbar. Initially the scroll bar is located at the highest point it can be.  

2) Place the mouse directly over the scroll bar and left click. Move mouse down enough so that text moves but **dont** let go of the mouse button. 

3) Keep the mouse at a fixed position with the left button fully pressed. The text should not be moving but it is.

If this isn't exactly or nearly what you are trying to show use similar step by step with explicit instructions to describe what you see.

**Comment by SlothLady** [comment]
Hiya, I just tried to reproduce it again and I have devised a way of reproducing it.


  1.  Start with a file with greater than 20649168 lines (exceeds address 013B14D0)
  2.  Scroll to address 013B14D0
  3.  Any further scrolling will result in the lines in the hex editor moving down the screen by 1px per scroll where before only the contents of the lines changed

I hope this makes a bit more sense, sorry again for my terrible explanation haha

**Comment by SlothLady** [comment]
Also when I say scrolling I explicitly mean with the scroll wheel on a mouse, not by moving the trackbar

**Comment by paxcut** [comment]
the problem is where to find such file? can you provide it? (drag it to the message)

**Comment by paxcut** [comment]
does it matter it file is all zeroes?

**Comment by SlothLady** [comment]
I shouldn't think it would matter

**Comment by SlothLady** [comment]
I think any file around 30mb should be enough

**Comment by paxcut** [comment]
Maybe this is related to how line numbers cant be bigger than 0x0FFFFFFF? but thats much bigger ~280 mbs

**Comment by paxcut** [comment]
you dont have to manually scroll all the lines from the top with the wheel do you?

**Comment by SlothLady** [comment]
No, I just scrolled to it with the scrollbar then used the scroll wheel after

**Comment by paxcut** [comment]
i used goto to put me at that location but scrolling from there seems normal.

**Comment by SlothLady** [comment]
![image](https://github.com/WerWolv/ImHex/assets/95983565/a2efdc90-86bd-43ee-9057-6bf2f580309a)
![image](https://github.com/WerWolv/ImHex/assets/95983565/562b19fc-4b5d-40a6-9253-ea78a6e636bb)

**Comment by paxcut** [comment]
To me, one wheel tick jumps about 5 lines. it always does that regardless of where it starts.

**Comment by SlothLady** [comment]
I'm not really sure how else to go about reproducing it, My screen resolution is 1920*1080, 100% scaling. This happens in windowed and fullscreen and occurs on version 1.30.1. I downloaded the portable version.

**Comment by paxcut** [comment]
Are you on windows?does the pixel scrolling continue no matter where you start after that location?

**Comment by SlothLady** [comment]
Yeah it happens no matter where I start after and yes I am on Windows 11

**Comment by paxcut** [comment]
Can you use pgup pgdn instead. or does it also act differently using the keyboard? also try the up/down arrows.

**Reply by rednnnno** [reply]
Funny I managed to reproduce it, a 80mb file about 1/2, included is a GIF; notice the top line offset changes. When at the beginning of the file this doesn't happen.

It's really not a huge issue but once you notice it you can't un-notice it (sorry for anyone readin this🤪)
![rolling](https://github.com/WerWolv/ImHex/assets/2566233/aab75485-ef10-4c0f-8eb8-eff684617f14)

**Comment by WerWolv** [comment]
This is due to how ImGui handles things internally and there's nothing I can do about it. (This is also the reason we have pages and not one continuous data page). 

In short, this happens due to floating point inaccuracies that start to happen more frequently the larger the numbers become. ImGui uses 32 bit floats internally to store the scroll / cursor offset. If it used doubles (or much better even, actual integers), this wouldn't happen as quickly (or at all)

Issue I opened about this ages ago: https://github.com/ocornut/imgui/issues/3609
